packages/marshal/src/types.js

```diff
@@ -2,68 +2,334 @@
+ *    * The `symbol` category includes only symbols which can be recreated
+ *      from string data. It therefore excludes JS unregistered symbols or
+ *      lisp/scheme gensyms. From JS it currently includes only
+ *      `Symbol.asyncIterator` (a so-called well-known symbol). From JS
+ *      it can include at most the well known symbols and the registered
+ *      symbols. (Because symbol registration names can collide with
+ *      well known symbol names, we will need another hilbert hotel
+ *      encoding to allow well known symbols and all registered symbols.)
+ *      From lisp, it can include symbols made from interned strings.
+ */
 
 /**
- * @typedef {*} Passable
+ * @typedef {Record<string | symbol, Function>} PrimaryRemotable
+ * An object marked with `Far` consisting of methods that can be remotely
+ * invoked. A PrimaryRemotable has a fresh unforgeable identity, i.e.,
+ * an unforgeable identity created when the object was created.
  *
- * A Passable value that may be marshalled. It is classified as one of
- * PassStyle. A Passable must be hardened.
+ * A PrimaryRemotable, being a kind of Remotable, is pass-by-remote. When a
+ * PrimaryRemotable T is passed as an argument in a message to an object in
+ * another vat, it arrives as a Remote<T>.
+ */
+
+/**
+ * Note on the odd use of `{_Remote: T}` below. Suggested at
+ * https://github.com/Agoric/agoric-sdk/pull/2909#discussion_r615522426 .
+ * We do not expect any `_Remote` property to actually exist. Since
+ * TS is structurally typed, we need some structure to make a distinct
+ * structure. IOW, using a non-existent property is a way to emulate
+ * nominal types in a structural type system.
  *
- * A Passable has a pass-by-copy superstructure. This includes the atomic
- * pass-by-copy primitives ("bigint" | "boolean" | "null" | "number" |
- * "string" | "undefined") and the composite pass-by-copy objects ("copyArray" |
- * "copyRecord" | "copyError"). The composite pass-by-copy objects that may
- * contain other Passables.
+ * TODO We can use this pseudo-property to obtain the `PrimaryRemotable` type.
+ * We can then extend the definition of `E()` to deal with `Remote<T>`s as if
+ * they have the methods and properties found on `T`.
  *
- * A Passable's pass-by-copy superstructure ends in PassableCap leaves. The
- * Passable can be further classified by the nature of these leaves. Since a
- * Passable is hardened, its structure and classification is stable --- its
- * structure and classification cannot change even if some of the objects are
- * proxies.
+ * @template {PrimaryRemotable} T
+ * @typedef {{_Remote: T}} Remote
+ * A `Remote<T>` is a Remotable object in one vat that represents a
+ * PrimaryRemotable of type T in another vat. The Remote has an unforgeable
+ * identity that is one-to-one with the unforgeable identity of its primary.
+ * If there is only one comm system between the two vats, then the identity of
```

Comment:
It wasn't immediately clear to me 1) why there would be more than one comm system (we don't have that in SwingSet, as vats are either connect via the kernel or the comms vat), and 2) what exactly the effect would be of two comms systems, and how that would even work. Maybe there is some context or background assumptions that could be added?